Mathlib/Combinatorics/Enumerative/Stirling.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2025 Beibei Xiong. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Beibei Xiong, Yu Shao, Weijie Jiang
+-/
+import Mathlib.Tactic.Ring
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Data.Nat.Choose.Basic
+
+/-!
+# Stirling Numbers
+
+This file defines Stirling numbers of the first and second kinds, proves their fundamental
+recurrence relations, and establishes some of their key properties and identities.
+-/
+
+/-!
+# The Stirling numbers of the first kind
+
+The unsigned Stirling numbers of the first kind, represent the number of ways
+to partition n distinct elements into k non-empty cycles.
+
+# The Stirling numbers of the second kind
+
+The Stirling numbers of the second kind, represent the number of ways to partition
+n distinct elements into k non-empty subsets.
+
+# Main definitions
+
+* `Nat.stirlingFirst`: the number of ways to partition n distinct elements into k non-empty
+  cycles, defined by the recursive relationship it satisfies.
+* `Nat.stirlingSecond`: the number of ways to partition n distinct elements into k non-empty
+  subsets, defined by the recursive relationship it satisfies.
+
+# References:
+
+* [Stirling Numbers of the First Kind – ProofWiki](https://proofwiki.org/wiki/Definition:Stirling_Numbers_of_the_First_Kind/Unsigned)
+* [Stirling Numbers of the Second Kind – ProofWiki](https://proofwiki.org/wiki/Definition:Stirling_Numbers_of_the_Second_Kind)
```

Comment:
I would rather add a standard reference such as Knuth, The Art of Computer Programming, volume 1, §1.2.